### PR TITLE
[py] Implement endpoint registration API on TChannel.

### PR DESCRIPTION
**Pull request by abhinav**
Given a service,

```thrift
// service.thrift

service Foo {
    i32 bar(1: string a, 2: list<bool> b) throws (1: SomethingWentWrong something);
}
```

You can do:

```python
from service import Foo

app = TChannel()

@app.register(Foo)
def bar(request, response, tchannel):
    a = request.args.a
    b = request.args.b

    if a == "hello":
        raise Foo.SomethingWentWrong()

    response.write_header("Timestamp", current_time())
    response.write_result(42)
```

Instead of ``write_result`` the handler can also just return the result (in a Future or directly).

```python
    response.write_header("Timestamp", current_time())
    return 42
```

For non-Thrift, you'll be able to do:

```python

@app.register('foo', 'json')
def foo(request, response, tchannel):
    # ...

@app.register('bar', 'raw')
def bar(request, response, tchannel):
    # ...

```

@blampe @breerly @junchaowu

### CODE REVIEW

File: python/tests/integration/trace/test_zipkin_trace.py [hunk 1]
```diff
@@ -57,24 +56,18 @@ def handler1(request, response, proxy):
...
 
 
 @pytest.yield_fixture
-def trace_server(random_open_port, handlers):
-    with TChannelServerManager(
-        port=random_open_port,
-        dispatcher=handlers
-    ) as manager:
+def trace_server(random_open_port):
+    with TChannelServerManager(random_open_port) as manager:
```
Comment (breerly) [inline python/tests/integration/trace/test_zipkin_trace.py]: You're shitting me right.

File: python/tests/integration/server_manager.py [hunk 2]
```diff
@@ -95,15 +109,13 @@ def shutdown(self):
 
 class TChannelServerManager(ServerManager):
```
Comment (breerly) [inline python/tests/integration/server_manager.py]: Rename to `TChannelTestServer` please. Remove hierarchy.

File: python/tchannel/tornado/tchannel.py [hunk 3]
```diff
@@ -51,6 +56,12 @@ class State(IntEnum):
 class TChannel(object):
     """Manages inbound and outbound connections to various hosts."""
 
+    _SCHEMES = {
+        'raw': scheme.RawArgScheme,
+        'json': scheme.JsonArgScheme,
```
Comment (junchaowu) [inline python/tchannel/tornado/tchannel.py]: Can we create a class containing fields to represent 'raw' , 'json' .. types?
I think ask use to type string input is dangerous. They may unsure it is capitalized or spell error.

Comment (abhinav) [inline python/tchannel/tornado/tchannel.py]: Yeah, I was gonna make this an enum or a set of pre-defined constants.

Comment (junchaowu) [inline python/tchannel/tornado/tchannel.py]: Actually I remember I had it in my retry patch. But it is quite out of date right now.

File: python/tests/integration/test_server.py [hunk 4]
```diff
@@ -25,46 +25,61 @@
...
+    def and_raise(self, exc):
 
-class ServerManager(object):
+        def execute(request, response):
+            raise exc
+
+        self.execute = execute
+
+
+class TestServer(object):
```
Comment (breerly) [inline python/tests/integration/test_server.py]: :clap: and the engineer thus managed to achieve code reuse without inheritance :clap:

Comment (abhinav) [inline python/tests/integration/test_server.py]: :shrug:

File: python/examples/handler.py [hunk 5]
```diff
@@ -58,19 +57,15 @@ def slow(request, response, proxy):
     response.flush()
 
 
-def get_example_handler():
-    dispatcher = RequestDispatcher()
+def register_example_endpoints(tchannel):
+    tchannel.register("hi", "raw", say_hi)
+    tchannel.register("ok", "raw", say_ok)
+    tchannel.register("echo", "raw", echo)
+    tchannel.register("slow", "raw", slow)
```
Comment (breerly) [inline python/examples/handler.py]: these might be more obvious if we use keyword args:

```
tchannel.register(endpoint="slow", argscheme="raw", handler=slow)
```

Comment (abhinav) [inline python/examples/handler.py]: Updated examples

**Comment by breerly** [comment]
:shipit:

**Pull request merged**